Library/OcCryptoLib/Sha2.c

```diff
@@ -583,11 +589,13 @@ Sha384Final (
   Context->Block[Context->Length] = 0x80;
   UNPACK64 (LenB, Context->Block + PmLen - 8);
 
-  Sha512Transform (Context, Context->Block, BlockNb);
-
-  for (Index = 0 ; Index < 6; ++Index) {
-    UNPACK64 (Context->State[Index], &HashDigest[Index << 3]);
+  if (mIsAvxEnabled) {
+    Sha512TransformAvx (Context->State, Context->Block, BlockNb);
+  } else {
+    Sha512Transform (Context, Context->Block, BlockNb);
   }
+
+  CopyMem (HashDigest, Context->State, SHA512_DIGEST_SIZE);
```

Comment:
I noticed there is a change in the logic for non-AVX version, is it still correct?

Reply:
According to APPENDIX C: SHA-512 EXAMPLES in FIPS Publication 180-2, non-AVX version became correct only after that change.

Reply:
Can confirm that it all works, going from about 8 secs to about 3 secs to run the 5,000,000 x password hash OC password verification.

As expected (presumably?) pre-existing password hash, salt no longer work, and have to be regenerated using new ocpasswordgen built against this changed version.

Reply:
Huh? How so? @mhaeuser was our previous implementation incorrect or what?

@MikeBeaton does Mac Secure Boot work?

Reply:
> @MikeBeaton does Mac Secure Boot work?

No, unfortunately not with this new implementation - and yes back on the previous one.

@MikhailKrichanov

Reply:
Is it confirmed it is *this* change that causes the issue?

Reply:
@MikhailKrichanov No, the original was correct.
> Throughout this specification, the “big-endian” convention is used when expressing both 32- and 64-bit words, so that within each word, the most significant bit is stored in the left-most bit position.

As calculations internally are done in little-endian order (e.g. you can see SHA*_H0 are stored word- and not byte-wise matching the spec pattern), the hash must be converted word-wise on finalisation. That is what the UNPACK code did.

Reply:
Can confirm that restoring the `UNPACK64` `for` loop, in the two places where this was changed to a `CopyMem`, changes back to working fine with previous password hashes and to passing Apple Secure Boot again.

cc @mhaeuser @vit9696 @MikhailKrichanov

Reply:
@MikhailKrichanov -

Some timings from the 'fixed' code (i.e. your PR, just slightly modified as above to restore `UNPACK64` in the two places where it was in fact required to match, e.g. Apple Secure Boot, previous OC passwords, etc.), compiled OC password verify times:

 - DEBUG: ASM: 5.3-5.4s C: 10.0-10.1s
 - RELEASE: ASM: 3.5-3.7s C: 4.8-4.9s
 
Just noting that while certainly not completely negligible, the speed up of ASM vs C in RELEASE is not _huge_ either... is this all in fact as expected? (Or am I - possibly - or the new code - possibly - doing something wrong?)

Cheers!

Reply:
Okay, so @vit9696 has mentioned that the existing speedup sounds about exactly right to him - and I gather @vit9696 and @mhaeuser both think that any additional remaining speedup to be gained would be in the shared code path (specifically `UNPACK64`) so it all sounds about right, speed-wise - many thanks!

Reply:
From a little more research, `UNPACK64` appears to be already optimised as a `bswap` in RELEASE build (only), so we may already (now, with this PR, that is!) be at peak SHA. 😊